File: 14.MultipleRowSelection/Sample Code/MultipleCheckmarksObjc/MultipleCheckmarksObjc/ViewController.m
#import "ViewController.h"

@interface ViewController () <UITableViewDataSource, UITableViewDelegate>

@property (weak, nonatomic) IBOutlet UITableView *tableView;
@property (nonatomic, strong) NSMutableArray *contents;
@property (nonatomic, strong) NSMutableArray *selectedIndexPaths;

@end

@implementation ViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    [self setContents:[@[ @"Red", @"Orange", @"Yellow", @"Green", @"Blue", @"Magenta"] mutableCopy]];
    [self setSelectedIndexPaths:[NSMutableArray array]];
}

- (IBAction)reloadButtonTapped:(UIButton *)sender
{
    [[self tableView] reloadData];
}

- (void)populateCell:(UITableViewCell *)cell
         atIndexPath:(NSIndexPath *)indexPath
{
    NSString *rowValue = [[self contents] objectAtIndex:[indexPath row]];
    
    [[cell textLabel] setText:rowValue];
    
    if ([[self selectedIndexPaths] containsObject:indexPath])
    {
        [cell setAccessoryType:UITableViewCellAccessoryCheckmark];
    }
    else
    {
        [cell setAccessoryType:UITableViewCellAccessoryNone];
    }
}

#pragma mark - UITableViewDataSource Methods

- (NSInteger)tableView:(UITableView *)tableView
 numberOfRowsInSection:(NSInteger)section
{
    return [[self contents] count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView
         cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *identifier = @"PlainCell";
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:identifier];
    if (cell == nil)
    {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:identifier];
    }
    
    [self populateCell:cell atIndexPath:indexPath];
    
    return cell;
}

#pragma mark - UITableViewDelegate Methods

- (void)tableView:(UITableView *)tableView
didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    
    if ([[self selectedIndexPaths] containsObject:indexPath])
    {
        [[self selectedIndexPaths] removeObject:indexPath];
    }
    else
    {
        [[self selectedIndexPaths] addObject:indexPath];
    }
    
    UITableViewCell *cell = [tableView cellForRowAtIndexPath:indexPath];
    [self populateCell:cell atIndexPath:indexPath];
}

@end

File: 08.BasicDataStructure/Sample Code/BasicDataStructureSwift/BasicDataStructureSwift/ViewController.swift
import UIKit

class ViewController: UIViewController, UITableViewDataSource, UITableViewDelegate
{
    
    @IBOutlet weak var tableView: UITableView!
    
    // MARK: - UITableViewDataSource Methods
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int
    {
        return 50
    }
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell
    {
        let identifier = "PlainCell"
        
        var cell = tableView.dequeueReusableCellWithIdentifier(identifier)
        if (cell == nil)
        {
            cell = UITableViewCell.init(style: .Default, reuseIdentifier: identifier)
        }
        
        if indexPath.row == 0
        {
            cell?.textLabel?.text = "Red"
        }
        else if indexPath.row == 1
        {
            cell?.textLabel?.text = "Orange"
        }
        else if indexPath.row == 2
        {
            cell?.textLabel?.text = "Yellow"
        }
        else if indexPath.row == 3
        {
            cell?.textLabel?.text = "Green"
        }
        else if indexPath.row == 4
        {
            cell?.textLabel?.text = "Blue"
        }
        else if indexPath.row == 5
        {
            cell?.textLabel?.text = "Magenta"
        }
        
        return cell!
    }
    
    // MARK: - UITableViewDelegate Methods
    
    func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath)
    {
        tableView.deselectRowAtIndexPath(indexPath, animated: true)
        
        if indexPath.row == 0
        {
            print("Red")
        }
        else if indexPath.row == 1
        {
            print("Orange")
        }
        else if indexPath.row == 2
        {
            print("Yellow")
        }
        else if indexPath.row == 3
        {
            print("Green")
        }
        else if indexPath.row == 4
        {
            print("Blue")
        }
        else if indexPath.row == 5
        {
            print("Magenta")
        }
    }
}

